### PR TITLE
trigger a custom JS event when hints are set (button ready)

### PR DESCRIPTION
**Pull request by danac-gs**
# Description
Show the PPC button disabled and enable it when the required data (hints with encrypted customer id)  is loaded. Prevent dead clicks and errors. Do not hardcode the solution (eg. CSS selectors), make it work via Magento configuration.

Solution: 
1. trigger a custom JS event from the core plugin code when data is loaded (ppc-hints-set)
2. through Magento config add "ppc-disabled" CSS class and JavaScript that listens to the aforementioned event and removes "ppc-disabled" class.

Magento Bolt Pay configuration:

**Additional Checkout Button Class**:
```
ppc-disabled
```
**Global CSS**:
```css
.bolt-product-checkout-button.ppc-disabled {
    opacity: 0.5;
    cursor: default;
    pointer-events: none;
}
```
**Additional Javascript**:
```javascript
$(document).on('ppc-hints-set', function(){
    $('.bolt-product-checkout-button').removeClass('ppc-disabled');
});
```
Fixes: https://app.asana.com/0/941920570700290/1163069420923242/f

(Optional: Add a changelog by writing a comment after "#changelog" on the next line)

#changelog Load hints data via promise, trigger a custom JS event when PPC required data is available, 

# Type of change

- [ ] Bug fix (change which fixes an issue)
- [x] New feature (change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?
Please validate that you have tested your change in at least one of the following areas:

- [ ] Successfully tested locally (or docker image)
- [x] Successfully tested on a staging or sandbox server
- [ ] Successfully tested on a merchant's staging server

# Checklist:

- [x] My code follows the style guidelines of this project.
- [x] I have performed a self-review of my own code.
- [ ] I have commented my code, particularly in hard-to-understand areas.
- [x] New and existing unit tests pass locally with my changes.
- [ ] I have created or modified unit tests to sufficiently cover my changes.
- [x] I have added my Asana task link and provided a changelog message if applicable.

### CODE REVIEW

**Comment by matt-thomason** [comment]
@danac-gs Does this mean every merchant who wants PPC needs to set this custom CSS and javascript? That seems really fragile. I know we are discouraging hard coding CSS and JS that manipulates the button, but this seems worse (making merchants write CSS and javascript that manipulates the button that we won't have good visibility on).

Did you talk with anyone from BX about how to slow-roll this button until hints are ready?
@daisy1754 @roopakv

**Comment by danac-gs** [comment]
@matt-thomason there are 2 opposing things to deal with here:

1. button show latency - show button as early as possible
2. dead clicks - prevent clicks until the checkout configuration is valid

similar to how is's ben done for the default checkout

I am refactoring it now to put the HTTP request in a promise. Still I'd leave this event dispatch in the core plugin. On slow connections (and complaint from the merchant) this might be useful adding to configuration.

The Magento default buttons are also shown disabled until ready

Moreover, the script requires 3 magento modules (loaded asynchronously) and if they are not initialized on button click the check returns false (dead click) and the checkout does not open.
They are used to verify

1. if all the product options are selected
2. is guest checkout allowed (for guest session)

we need to know that before we can let the modal open

File: view/frontend/templates/button_product_page.phtml [hunk 1]
```diff
@@ -207,31 +202,35 @@ if (!$block->isSupportableType()) return;
...
+                    $.get(settings.get_hints_url)
+                        .done(function(data) {
+                            hints = data.hints;
+                        })
+                        .fail(function() {
+                            hints = {prefill:{}};
+                        })
+                        .always(function() {
+                            resolve(hints);
+                            // disapatch event when required data is ready
```
Comment (vitaliyreznikov) [inline view/frontend/templates/button_product_page.phtml]: typo in word dispatch

File: view/frontend/templates/button_product_page.phtml [hunk 2]
```diff
@@ -207,31 +202,35 @@ if (!$block->isSupportableType()) return;
                         options: JSON.stringify(options)
                     }]
                 };
+
+                var hintsPromise = new Promise(function (resolve, reject) {
+                    if (hints) {
+                        resolve(hints);
+                        return;
+                    }
+                    $.get(settings.get_hints_url)
```
Comment (vitaliyreznikov) [inline view/frontend/templates/button_product_page.phtml]: Do you think we can (in separate PR) find a way how to call the server only if a user logged in?

File: view/frontend/templates/button_product_page.phtml [hunk 3]
```diff
@@ -207,31 +202,35 @@ if (!$block->isSupportableType()) return;
                         options: JSON.stringify(options)
                     }]
                 };
+
+                var hintsPromise = new Promise(function (resolve, reject) {
```
Comment (vitaliyreznikov) [inline view/frontend/templates/button_product_page.phtml]: Maybe we have a little problem. If we hints ajax call works long enough (10 seconds), user can click product options, configureProductCheckout calls again, hints call again.
Bolt modal wait for last ajax call and can as result show timeout error.
Maybe it's better to create promise only once, not on each setupProductPage() call

Comment (matt-thomason) [reply view/frontend/templates/button_product_page.phtml]: @vitaliyreznikov Can you create a backlog task in the M2 asana board to investigate the identified corner case?

File: view/frontend/templates/button_product_page.phtml (hunk 2 above)
Comment (vitaliyreznikov) [reply view/frontend/templates/button_product_page.phtml]: @danac-gs I'd like to create asana about it if you don't mind.

File: view/frontend/templates/button_product_page.phtml (hunk 3 above)
Comment (vitaliyreznikov) [reply view/frontend/templates/button_product_page.phtml]: @matt-thomason Sure. https://app.asana.com/0/941920570700290/1165280808501785/f

**Pull request merged**